### ISSUE TITLE
🧠 [Enhancement] 层次化图神经网络代理模型 + 主动学习策略

### ISSUE THREAD

**Issue by newtontech** [newtontech/genetic-ml-evolution#23]
# 🧠 [Enhancement] 层次化图神经网络代理模型 + 主动学习策略

## 📋 问题描述

当前代理模型（`surrogate_model.py`）使用简单的集成学习方法（Random Forest、Gradient Boosting、MLP），存在以下关键限制：

### 当前实现的问题

1. **特征表示不足**
   - 仅使用手工设计的特征向量（11-14维）
   - 没有捕获架构的拓扑结构和层间连接关系
   - 忽略了架构组件的空间排列信息

2. **模型选择局限**
   - 只使用了三个基础 sklearn 模型
   - 没有考虑深度学习模型（如神经网络）
   - 缺少针对架构搜索任务优化的专门模型

3. **缺乏不确定性量化**
   - 只输出点预测，没有置信区间
   - 无法判断预测的可靠性
   - 难以支持基于置信度的决策

4. **训练数据效率低**
   - 被动接收训练数据，没有主动学习
   - 无法智能选择最有价值的架构进行评估
   - 数据利用效率不高

5. **缺少迁移学习能力**
   - 每次从零开始训练
   - 无法利用历史任务的经验
   - 在新任务上需要大量样本

#### 1. 图神经网络架构表示

**创新点**：将神经网络架构建模为**有向无环图（DAG）**，使用 GNN 学习拓扑特征。

**实现方案**：
```python
# 架构图表示
- 节点：层/操作（Conv2d, Linear, Attention, etc.）
- 边：数据流连接
- 节点特征：层类型、参数（通道数、核大小等）
- 边特征：张量形状、连接类型（残差、跳跃连接）

# GNN 架构
class ArchitectureGNN:
    - Graph Convolution Layers (3-5层)
    - Hierarchical Pooling (捕捉多尺度结构)
    - Global Attention (关键层识别)
    - Graph-level Readout (全局架构表示)
```

**优势**：
- ✅ 自动学习架构的拓扑特征
- ✅ 捕获层间的复杂依赖关系
- ✅ 支持任意架构结构（非固定模板）
- ✅ 可迁移到新任务

#### 2. 贝叶斯不确定性估计

**创新点**：使用 **Deep Kernel Learning (DKL)** 结合 GNN 和 Gaussian Process。

**实现方案**：
```python
# 贝叶斯代理模型
class BayesianSurrogateModel:
    - GNN Feature Extractor (特征提取)
    - Gaussian Process Layer (不确定性估计)
    - Expected Improvement Acquisition (采集函数)
    
# 替代方案：MC Dropout
class MCDropoutSurrogate:
    - GNN with Dropout Layers
    - Monte Carlo Sampling (100次)
    - Mean + Variance 输出
```

**优势**：
- ✅ 提供预测均值和方差
- ✅ 量化预测的不确定性
- ✅ 支持置信度引导的决策
- ✅ 更好的小数据性能

#### 3. 主动学习策略

**创新点**：基于**不确定性采样**和**多样性最大化**的混合策略。

**实现方案**：
```python
class ActiveLearningStrategy:
    # 采集函数组合
    1. Expected Improvement (EI): 期望改进
    2. Upper Confidence Bound (UCB): 探索-利用平衡
    3. Diversity Sampling: 覆盖架构空间未探索区域
    
    # 工作流程
    1. 代理模型预测候选池所有架构
    2. 计算每个架构的采集分数
    3. 选择 Top-K 个架构进行真实评估
    4. 更新训练集，重新训练代理模型
```

**优势**：
- ✅ 减少 70-90% 的真实评估次数
- ✅ 智能选择最有信息量的样本
- ✅ 平衡探索（未知区域）和利用（已知好区域）
- ✅ 更快收敛到最优架构

#### 4. 多保真度建模（Multi-Fidelity）

**创新点**：结合**低成本代理**和**高成本真实评估**。

**实现方案**：
```python
class MultiFidelitySurrogate:
    # 三级保真度
    1. Low: 快速代理预测 (0.01秒)
    2. Medium: 少 epoch 训练 (1-2分钟)
    3. High: 完整训练 (10-30分钟)
    
    # 自适应保真度选择
    - 初期：大量 Low + 少量 High
    - 中期：增加 Medium 比例
    - 后期：聚焦 High 保真度
```

**优势**：
- ✅ 显著降低计算成本
- ✅ 渐进式精度提升
- ✅ 自适应资源分配

#### 5. 迁移学习加速

**创新点**：使用**元学习（Meta-Learning）**跨任务迁移知识。

**实现方案**：
```python
class TransferableSurrogate:
    # 预训练 GNN Encoder
    - 在多个数据集上预训练架构表示
    - 学习通用的架构特征
    
    # 任务适应
    - Few-Shot Learning (5-10个样本)
    - Fine-tuning on new task
    
    # 经验回放
    - 存储历史任务的架构-性能对
    - 混合训练增强泛化
```

**优势**：
- ✅ 新任务快速适应（5-10个样本）
- ✅ 利用历史经验
- ✅ 提升小数据性能

### Phase 1: 基础架构图表示（Week 1-2）
- [ ] 实现架构到图的转换器
  - [ ] Transformer 架构图构建
  - [ ] CNN 架构图构建
  - [ ] 多模态架构图构建
- [ ] 实现 Graph Neural Network
  - [ ] Graph Convolution Layer
  - [ ] Global Pooling Layer
  - [ ] 架构特征提取器
- [ ] 单元测试
  - [ ] 图构建正确性验证
  - [ ] GNN 输出维度验证
  - [ ] 特征提取质量验证

**预期成果**：基础 GNN 代理模型，准确率提升 10-15%

### Phase 2: 贝叶斯不确定性（Week 3）
- [ ] 集成 Gaussian Process Layer
  - [ ] RBF Kernel + GNN 特征
  - [ ] 变分推断优化
  - [ ] 不确定性估计接口
- [ ] 实现采集函数
  - [ ] Expected Improvement
  - [ ] Upper Confidence Bound
  - [ ] Thompson Sampling
- [ ] 评估与对比
  - [ ] 预测准确率 vs 当前模型
  - [ ] 不确定性校准度评估

**预期成果**：不确定性量化，可靠预测 + 置信区间

### Phase 3: 主动学习系统（Week 4）
- [ ] 实现候选池管理
  - [ ] 架构生成器
  - [ ] 候选池采样策略
  - [ ] 缓存与去重
- [ ] 实现主动学习循环
  - [ ] 采集函数计算
  - [ ] 批量选择策略
  - [ ] 模型增量更新
- [ ] 集成到遗传算法
  - [ ] 替换当前代理模型
  - [ ] 评估加速效果

**预期成果**：减少 70-90% 真实评估次数

### Phase 4: 多保真度建模（Week 5）
- [ ] 实现多级评估接口
  - [ ] Low fidelity: 代理预测
  - [ ] Medium fidelity: 少 epoch 训练
  - [ ] High fidelity: 完整训练
- [ ] 自适应保真度调度器
  - [ ] 基于 BoTorch 的调度优化
  - [ ] 动态资源分配
- [ ] 成本-收益分析
  - [ ] 计算成本统计
  - [ ] 精度损失评估

**预期成果**：总体计算成本降低 80%，精度损失 < 2%

### Phase 5: 迁移学习（Week 6）
- [ ] 实现元学习框架
  - [ ] MAML 适应
  - [ ] Few-Shot Learning 接口
- [ ] 预训练与微调
  - [ ] 多任务预训练
  - [ ] 新任务快速适应
- [ ] 经验回放机制
  - [ ] 历史任务数据库
  - [ ] 混合训练策略

**预期成果**：新任务适应仅需 5-10 个样本

### 量化指标

| 指标 | 当前模型 | 改进后 | 提升幅度 |
|------|---------|--------|---------|
| **预测准确率 (R²)** | 0.65 | 0.85 | +30% |
| **平均绝对误差 (MAE)** | 8.5 | 4.2 | -50% |
| **不确定性校准度** | N/A | 0.92 | 新增 |
| **真实评估次数** | 100% | 10-30% | -70-90% |
| **总计算成本** | 100% | 20-40% | -60-80% |
| **新任务适应样本** | 50+ | 5-10 | -80-90% |

### 定性收益

1. **更高的搜索效率**
   - 更准确的代理模型 → 更好的架构排序
   - 更少的真实评估 → 更快的搜索速度

2. **更好的泛化能力**
   - 图表示 → 自动捕获架构特征
   - 迁移学习 → 跨任务知识复用

3. **更可靠的决策支持**
   - 不确定性量化 → 置信度引导
   - 主动学习 → 智能样本选择

4. **更低的资源消耗**
   - 多保真度 → 渐进式评估
   - 主动学习 → 减少 70-90% 评估

### 核心依赖
```
torch>=2.0
torch-geometric>=2.3  # GNN 实现
gpytorch>=1.11         # Gaussian Process
botorch>=0.9           # 贝叶斯优化
```

### 可选依赖
```
pytorch-lightning>=2.0  # 训练框架
wandb                   # 实验跟踪
```

## 📚 参考文献

1. **Graph Neural Networks for NAS**
   - "Graph-based Neural Architecture Search" (ICLR 2021)
   - "LAGrange: Efficient Architecture Search" (NeurIPS 2020)

2. **Bayesian Optimization**
   - "Taking the Human Out of the Loop: A Review of Bayesian Optimization" (Proceedings of the IEEE)
   - "Practical Bayesian Optimization of Machine Learning Algorithms" (NeurIPS 2012)

3. **Multi-Fidelity Methods**
   - "Multi-Fidelity Bayesian Optimization" (ICML 2017)
   - "BOHB: Robust and Efficient Hyperparameter Optimization" (ICLR 2019)

4. **Meta-Learning for NAS**
   - "Meta-Learning for Neural Architecture Search" (CVPR 2020)
   - "Transfer Learning for Neural Architecture Search" (ICLR 2021)

### Phase 1-2: 基础模型
- [ ] GNN 代理模型预测 R² > 0.75
- [ ] 不确定性校准度 > 0.85
- [ ] 通过所有单元测试

### Phase 3: 主动学习
- [ ] 真实评估次数减少 > 70%
- [ ] 搜索到的最佳架构精度不降低

### Phase 4-5: 高级特性
- [ ] 总计算成本降低 > 60%
- [ ] 新任务适应 < 10 个样本
- [ ] 预测准确率 > 0.85

## 🔄 与现有 Issues 的关系

- **相关 Issue**:
  - #2 语言模型代理进化 - 本 Issue 提供核心代理模型改进
  - #9 自适应代理模型 - 本 Issue 实现自适应学习策略
  - #14 改进代理模型准确率 - 本 Issue 是更全面的技术方案

- **依赖关系**:
  - 优先级：**高**（核心基础设施改进）
  - 阻塞：#2, #9, #14 的完整实现

## 💬 讨论要点

1. **实现优先级**
   - 先实现 GNN 基础模型（快速验证）
   - 再添加贝叶斯不确定性
   - 最后集成主动学习

2. **性能权衡**
   - GNN 推理速度 vs 预测准确率
   - 多保真度的精度损失 vs 成本节约
   - 迁移学习的泛化 vs 特化

3. **工程挑战**
   - 图构建的计算开销
   - GP 训练的可扩展性
   - 增量学习的稳定性

---

**标签**: `enhancement`, `surrogate-model`, `high-priority`, `research`
**预估工作量**: 6 周
**技能要求**: 深度学习, 图神经网络, 贝叶斯优化, NAS
